SX-2129 Edit phone number Circuit UI component

roma-claudio opened the pull request:
Addresses SX-2129

## Purpose

Update edit phone number component

## Approach and changes

1. Improved test coverage for `PhoneNumberInput.tsx` and `PhoneNumberInput.service.tsx`
2. Granular control of `readonly` and `invalid` attributes for each field (country code and subscriber number)
3. Updated country code field pattern to limit the number of max allowed digits
4. Added `placeholder` prop to the subscriber number
5. Updated docs

## Definition of done

* [ ] Development completed
* [ ] Reviewers assigned
* [ ] Unit and integration tests
* [ ] Meets minimum browser support
* [ ] Meets accessibility requirements

roma-claudio commented on packages/circuit-ui/components/PhoneNumberInput/PhoneNumberInputService.ts:
```diff
@@ -33,7 +33,8 @@ export function mapCountryCodeOptions(
   locale?: string | string[],
 ): Required<SelectProps>['options'] {
-  // TODO: Check whether Intl.DisplayNames is supported and add fallback logic
+  // TODO Intl is only unsupported in Opera 50, iOS Safari 11.0-11.2, Firefox 67, Edge 79, Chrome 63. All of those version are not supported by Circuit UI, so should we even care?
```
Do we have to make sure that the code works for unsupported browser versions?

connor-baer replied:
What are these support stats based on? 

According to caniuse.com, `Intl.DisplayNames` is [not supported](https://caniuse.com/mdn-javascript_builtins_intl_displaynames) in Chrome <81, Safari <14.1, Firefox <86 and Opera <68.

roma-claudio replied:
Correct, I've only checked the Intl object instead of the Intl.DisplayNames method, apparently they differ in terms of supported browsers 👍🏻

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Intl
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Intl/DisplayNames

roma-claudio replied:
Code updated, the fallback case will use the list of countries passed as keys

connor-baer commented on packages/circuit-ui/components/PhoneNumberInput/PhoneNumberInput.tsx:
```diff
@@ -254,27 +295,31 @@ export const PhoneNumberInput = forwardRef<
             hideLabel
             autoComplete="tel-country-code"
-            invalid={invalid}
+            invalid={invalid || countryCode.invalid}
             disabled={disabled}
             className={classes['country-code']}
+            aria-readonly={readOnly || countryCode.readonly}
             {...countryCode}
```
The `invalid` and `aria-readonly` props will be overwritten by the spread `{...countryCode}` props below. Move them below to avoid this.

roma-claudio replied:
Code updated

connor-baer commented on packages/circuit-ui/components/PhoneNumberInput/PhoneNumberInput.tsx:
```diff
@@ -254,27 +295,31 @@ export const PhoneNumberInput = forwardRef<
-            placeholder="123 456789"
-            pattern="[0-9]+[0-9 ]+"
+            placeholder={subscriberNumber.placeholder}
+            pattern="^(?:[0-9]\s?){0,13}[0-9]$"
```
The maximum number of digits is 15, isn't it? So it should be 14 digits with an optional trailing space + 1 final digit.
```suggestion
            pattern="^(?:[0-9]\s?){0,14}[0-9]$"
```

roma-claudio replied:
Correct, I was (wrongly) assuming that the max number of digits of the format applied both to country code and subscriber number, but it applies only to the subscriber number 👍🏻

connor-baer commented on packages/circuit-ui/components/PhoneNumberInput/PhoneNumberInputService.spec.ts:
```diff
@@ -63,10 +63,40 @@ describe('PhoneNumberInputService', () => {
+      expect(actual[1].label).toBe('United States (+1)');
+    });
+
+    it('Shold use the locale as the default country code', () => {
```
```suggestion
    it('should use the locale as the default country code', () => {
```

connor-baer commented on packages/circuit-ui/components/PhoneNumberInput/PhoneNumberInput.mdx:
```diff
@@ -7,11 +7,79 @@ import * as Stories from './PhoneNumberInput.stories';
+
+- **Do** always accompany the `PhoneNumberInput` field with a visible label. This helps users understand what information is required.
+- **Do not** use `PhoneNumberInput` for non-phone number data. It is specifically designed for phone number input and includes specific formatting and validation for phone numbers.
+- **Do not** communicate that a `PhoneNumberInput` field is required on the label or the placeholder. We assume that everything is required for the user unless explicitly said so.
```
```suggestion
- **Do not** communicate that a `PhoneNumberInput` field is required on the label or the placeholder. We assume that everything is required for the user unless explicitly stated otherwise.
```

connor-baer commented on packages/circuit-ui/components/PhoneNumberInput/PhoneNumberInput.mdx:
```diff
@@ -7,11 +7,79 @@ import * as Stories from './PhoneNumberInput.stories';
+
+## With Prefix
+
+The `PhoneNumberInput` component supports prefixes. Use the `withPrefix` prop to add a small prefix to the country code selected input field. This can be useful for adding a flag icon, which can help the user recognize the selected country.
+
+- **Do** use the `withPrefix` prop to provide relevant visual cues that help guide the user's input.
```
```suggestion
The `PhoneNumberInput` component supports prefixes. Use the `renderPrefix` prop to add a small prefix to the country code selected input field. This can be useful for adding a flag icon, which can help the user recognize the selected country.

- **Do** use the `renderPrefix` prop to provide relevant visual cues that help guide the user's input.
```

connor-baer commented on packages/circuit-ui/components/PhoneNumberInput/PhoneNumberInput.stories.tsx:
```diff
@@ -35,13 +48,99 @@ Base.args = {
   validationHint: 'Maximum 15 digits',
 };
+
+export const Invalid = (args: PhoneNumberInputProps) => (
+  <PhoneNumberInput {...args} />
+);
+
+Invalid.args = {
+  ...Base.args,
+  validationHint: 'This number is not valid',
+  subscriberNumber: {
+    ...Base.args.subscriberNumber,
+    invalid: true,
+  },
+};
+
+export const Warning = (args: PhoneNumberInputProps) => (
+  <PhoneNumberInput {...args} />
+);
+
+Warning.args = {
+  ...Base.args,
+  hasWarning: true,
+  validationHint: 'This phone number may incur additional charges',
+};
+
+export const Valid = (args: PhoneNumberInputProps) => (
+  <PhoneNumberInput {...args} />
+);
+
+Valid.args = {
+  ...Base.args,
+  showValid: true,
+  validationHint: 'This phone number looks good',
+};
```
We use chromatic.com to catch visual bugs by taking a screenshot of every story. We're billed per story, so we try to combine related stories into one to save costs. 

Could you please merge these into a single "Validations" story, similar to the Input component?

roma-claudio replied:
Code updated

connor-baer commented on packages/circuit-ui/components/PhoneNumberInput/PhoneNumberInput.tsx:
```diff
@@ -102,51 +103,90 @@ export interface PhoneNumberInputProps
+     * Triggers readonly styles on the component.
      */
     readonly?: boolean;
+    /**
+     * Callback when the country code changes.
+     * The callback receives the normalized phone number in the [E.164 format](https://en.wikipedia.org/wiki/E.164),
+     * e.g. `+17024181234`.
+     */
```
This is inaccurate. This `onChange` callback is specific to the country code `select` element and receives the change `event` (which only contains the country code as the `event.target.value`).

connor-baer commented on packages/circuit-ui/components/PhoneNumberInput/PhoneNumberInput.tsx:
```diff
@@ -102,51 +103,90 @@ export interface PhoneNumberInputProps
+     * Triggers readonly styles on the component.
+     */
+    readonly?: boolean;
+    /**
+     * Callback when the subscriber number changes.
+     * The callback receives the normalized phone number in the [E.164 format](https://en.wikipedia.org/wiki/E.164),
+     * e.g. `+17024181234`.
+     */
```
Same here. This `onChange` callback is specific to the subscriber number `input` element and receives the change `event` (which only contains the raw subscriber number as the `event.target.value`).

roma-claudio replied:
Correct, code updated

connor-baer reviewed (commented):
Thank you for finishing the component and especially for writing such detailed docs. Great job! 🌟  

I don't see any logic to handle pasting of a (full) phone number. Have you tested how the component could/would handle this?

roma-claudio commented on packages/circuit-ui/components/PhoneNumberInput/PhoneNumberInput.module.css:
```diff
@@ -7,13 +7,24 @@
 }
 
+.country-code[aria-readonly="true"] {
+  pointer-events: none;
```
Added a logic to set a negative tabindex when the readonly prop is truthy

roma-claudio replied:
Sounds like a good idea! I've adapted the code to implement this logic. I had to do few typescript casts but all in all it's a relatively clean implementation, let me know if there's anything concerning you 👍🏻

connor-baer replied:
It looks like you forgot to remove the `useEffect` that sets the `tabindex`. This isn't needed anymore, is it?

roma-claudio replied:
Correct, code updated

roma-claudio commented:
> Thank you for finishing the component and especially for writing such detailed docs. Great job! 🌟
> 
> I don't see any logic to handle pasting of a (full) phone number. Have you tested how the component could/would handle this?

At the moment we don't handle this scenario, the question is: how do we want to handle it? Stripping it and setting the country code and subscriber number values looks error prone in this case because those can have different lengths and it's difficult to determine them 🤔

connor-baer reviewed (commented):
> At the moment we don't handle this scenario (meaning the entire phone number will be pasted in the subscriber number field), the question is: how do we want to handle it? Stripping it and setting the country code and subscriber number values looks error prone in this case because those can have different lengths and it's difficult to determine them 🤔

I looked into this when I initially researched this component. It's possible to reliably extract the country code from a complete phone number, assuming that all country codes are known. Given that the component usually only has access to a subset of country codes (from the `options` prop), we will only be able to match supported country codes 🤔 I think that's fine though since I would consider unsupported country codes an edge case.

I'm happy to tackle this in a follow-up PR. I don't think it's necessary for the MVP.

roma-claudio commented on packages/circuit-ui/components/PhoneNumberInput/PhoneNumberInput.tsx:
```diff
@@ -254,42 +301,44 @@ export const PhoneNumberInput = forwardRef<
-          hasWarning={hasWarning}
-          showValid={showValid}
-          validationHint={validationHint}
-        />
```
Good catch, I moved it back as it was 👍🏻

roma-claudio commented:
> > At the moment we don't handle this scenario (meaning the entire phone number will be pasted in the subscriber number field), the question is: how do we want to handle it? Stripping it and setting the country code and subscriber number values looks error prone in this case because those can have different lengths and it's difficult to determine them 🤔
> 
> I looked into this when I initially researched this component. It's possible to reliably extract the country code from a complete phone number, assuming that all country codes are known. Given that the component usually only has access to a subset of country codes (from the `options` prop), we will only be able to match supported country codes 🤔 I think that's fine though since I would consider unsupported country codes an edge case.
> 
> I'm happy to tackle this in a follow-up PR. I don't think it's necessary for the MVP.

Ah ok I now see your point, indeed the country code can be extracted from a full number if: 
1. The list of country codes is known
2. There are no ambiguous country codes, e.g. +20 and +201, but I've checked the list of country codes and this is never the case: https://en.wikipedia.org/wiki/List_of_country_calling_codes )

Unfortunately I've no capacity to address this right now, so if you agree let's have it in a follow-up PR

roma-claudio commented:
> Thank you for addressing all the feedback. This looks great!
> 
> Let's merge it back into #2500, then I'll implement the logic for pasting a phone number so we can release the component 🎉

Great! I'll merge it right away then, do you have a rough idea on when we can release the component?

connor-baer commented:
> do you have a rough idea on when we can release the component?

I'm aiming for early next week 🤞🏻
